lib/Parse/ParseExpr.cpp

```diff
@@ -2110,6 +2113,56 @@ ParserResult<Expr> Parser::parseExprStringLiteral() {
                                       AppendingExpr));
 }
 
+ParserResult<Expr> Parser::parseExprRegexLiteral() {
+  assert(Tok.is(tok::regex_literal));
+  SyntaxParsingContext LocalContext(SyntaxContext,
+                                    SyntaxKind::RegexLiteralExpr);
+
+  auto *parseRegexFn = Lexer::getExperimentalRegexParsingFn();
+  assert(parseRegexFn);
+
+  // Strip off delimiters.
+  auto rawText = Tok.getText();
+  assert(rawText.front() == '\'' && rawText.back() == '\'');
+  auto regexText = rawText.slice(1, rawText.size() - 1);
+
+  // Let the Swift library parse the contents, returning an error, or null if
+  // successful.
+  // TODO: We need to be able to pass back a source location to emit the error
+  // at.
+  auto *errorStr = parseRegexFn(regexText.str().c_str());
+  if (errorStr)
+    diagnose(Tok, diag::regex_literal_parsing_error, errorStr);
+
+  auto loc = consumeToken();
+
+  // Create an implicit .init(_regexString: "<regex text>") call to serve as the
+  // semantic expression for the regex. The type-checker will provide it with
+  // the correct contextual type. We delay the contextual type for a couple of
+  // reasons:
+  // 1. We need to delay type lookup until after parsing.
```

Comment:
Eventually (hopefully soon) the parser will also communicate back info about the capture structures. Would this approach extend naturally to handle that?

Reply:
I think that would work fine, we could attach that info to the AST and resolve it when type-checking.